### ISSUE TITLE
[CHILD] Update lane model

### ISSUE THREAD

**Issue by reomelo** [SEAME-pt/HotWheels-Lane_Detection#61]
`Description:` The team will be divided into 2 groups. Each group will change the model at will and in the end we'll make a comparison to choose which one to use.

</br>

`Definition of done:` The issue is considered as done once we have a new updated model that shows improvement when compared to the one we have now and provides satisfactory results. IT SHOULD WORK IN CARLA AND IN THE LAB!

**Comment by Fle-bihh** [comment]
First tested solution will be : Apply fine-tuning to the current lane-model using a sub-part of the CurveLane dataset

**Comment by Fle-bihh** [comment]
We need to determine the limit layer for fine-tuning. Every layers onward would be unfreeze (meaning their weight will be available for training again).

**Comment by Fle-bihh** [comment]
The model uses a U-Net style encoder-decoder with 3 encoder blocks (encoder_conv1_*, encoder_conv2_*, encoder_conv3_*) and a bottleneck (middle_conv*).
Since early layers capture general features (edges, gradients) while deeper layers learn task-specific patterns (lane shapes), I will unfreeze from encoder_conv2_* onwards.
This should allow the model to adapt higher-level features to better handle curves (underrepresented in CULane but present in CurvesLane) without disrupting low-level feature extraction.

**Comment by Fle-bihh** [comment]
Depending on duration of training, I should also try unfreezing from encoder_conv_1 and compare results.

**Comment by Fle-bihh** [comment]
CurveLanes dataset labels format and location differs from the CULane one. Realized a new script to preprocess the data and get it in the format that our model expects

**Comment by Fle-bihh** [comment]
Currently Downloading the CurveLanes dataset (more than gb). As soon as it is done, the first fine-tuning will start with following layers's state :

`🟢 Trainable layers:
model_input: frozen
gaussian_blur: frozen
sobel_x: frozen
sobel_y: frozen
gradient_magnitude: frozen
non_max_suppression: frozen
adaptive_threshold: frozen
encoder_conv1_1: frozen
batch_normalization: frozen
encoder_conv1_2: frozen
batch_normalization_1: frozen
encoder_pool1: frozen
spatial_dropout2d: frozen
encoder_conv2_1: trainable
batch_normalization_2: trainable
encoder_conv2_2: trainable
batch_normalization_3: trainable
encoder_pool2: trainable
spatial_dropout2d_1: trainable
encoder_conv3_1: trainable
batch_normalization_4: trainable
encoder_conv3_2: trainable
batch_normalization_5: trainable
encoder_pool3: trainable
spatial_dropout2d_2: trainable
middle_conv1: trainable
batch_normalization_6: trainable
middle_conv2: trainable
batch_normalization_7: trainable
spatial_dropout2d_3: trainable
decoder_up3: trainable
conv2d: trainable
concatenate: trainable
decoder_conv3_1: trainable
batch_normalization_8: trainable
decoder_conv3_2: trainable
batch_normalization_9: trainable
decoder_up2: trainable
conv2d_1: trainable
concatenate_1: trainable
decoder_conv2_1: trainable
batch_normalization_10: trainable
decoder_conv2_2: trainable
batch_normalization_11: trainable
decoder_up1: trainable
conv2d_2: trainable
concatenate_2: trainable
decoder_conv1_1: trainable
batch_normalization_12: trainable
decoder_conv1_2: trainable
batch_normalization_13: trainable
lane_attention: trainable
output: trainable`

**Comment by MicchelFAB** [comment]
## New Dataset to train
[CurveLanes Dataset](https://github.com/SoulmateB/CurveLanes?tab=readme-ov-file)`

#### Downloading and extracting on the `~/Curvelanes` folder, in progress.

**Comment by Fle-bihh** [comment]
All dataset is downloaded. Script made to preprocess the data into the expected model's input done. will start training with 25% of the data.

**Comment by Fle-bihh** [comment]
Currently training with following parameters :

Dataset: CurvesLane (25% of training set)

Epochs: 10

Unfreeze from: encoder_conv2_*

Learning rate: original LR / 10

Output_dir : output-20250424-121511

Estimated duration : 3 hours

**Comment by Fle-bihh** [comment]
Training results :
📋 Final loss: 0.2829
📋 Final accuracy: 0.9800
📋 Final IoU: 0.3810

**Comment by Fle-bihh** [comment]
The new model has been tested on Carla simulator and results are more than satisfying, every curves are appearing on the mask and even straight lines seem to be perfected better.

**Comment by Fle-bihh** [comment]
The new model has now also been tested and the results are not as satisfying. Definitely an improvement but would need some training on lab data now to get perfect accuracy.

**Comment by Fle-bihh** [comment]
I will now start working on fine-tuning the current fine-tuned model with lab images. For this I have two solutions :
1 - Use the dataset that team 02 generated, which is mix of lab images and Carla images (mostly Carla images) and for which they already hand-made the labels.
2 - Use the script from team02 to generate labels on our lab images and then train on it. 

Since our results in Carla are almost perfect when the car is on a lane, but still lacking quality at intersections (cross-walks etc...), I think training on a dataset containing Carla images would not be useless. And it would also be time-saving to use the dataset already made instead of generating new labels myself.

So I will use the dataset from team02, and go for solution 2 only if we still lack perception of lab lanes after applying solution 1. In both cases, data augmentation will be applied to the dataset before training to improve even more the quality of prediction in any kind of scenarios.

**Comment by Fle-bihh** [comment]
Applying on-the-fly data augmentation (random rotations, flips, brightness/contrast adjustments, Gaussian noise, and zoom) during training—this does not physically expand the dataset on disk, but generates varied versions of each sample each epoch.

**Comment by Fle-bihh** [comment]
Currently training with following parameters :

Dataset: Team02 Dataset (710 images)

Epochs: 20

Unfreeze encoder_conv1_, lane_attention + output

Learning rate: original LR / 10

Output_dir : output-20250428-123750

Estimated duration : 1 hours

Initial loss: 0.6329 | acc: 0.9696 | IoU: 0.0719

**Comment by Fle-bihh** [comment]
After training : Final loss: 0.5554 | acc: 0.9721 | IoU: 0.1444
I will try the new model on the car but the training might need way more epochs.

**Comment by Fle-bihh** [comment]
Actually the training stopped at epoch 12/20 because of the realy-stopping callback. But this callback was monitoring the val_loss, and it would be more accurate if monitoring val_binary_mean_iou. So i updated the create_callbacks method and will start the training again.

**Comment by Fle-bihh** [comment]
Currently training with following parameters :

Dataset: Team02 Dataset (710 images)

Epochs: 50

Unfreeze encoder_conv1_, lane_attention + output

Learning rate: original LR / 10

Output_dir : output-20250428-130242

Estimated duration : 50 minutes

EarlyStopping callback monitoring : val_binary_mean_iou

Initial loss: 0.6329 | acc: 0.9696 | IoU: 0.0719

**Comment by Fle-bihh** [comment]
This time it stopped at epoch 18/50 (Meaning it wasnt learning anymore)

Final loss: 0.5267 | acc: 0.9727 | IoU: 0.1665

Will now test it on the car

**Comment by Fle-bihh** [comment]
After trying on Carla sim, the model did get "worst". So I need to change the layers I unfroze for the training and train again.

**Comment by Fle-bihh** [comment]
This time, unfreezing only the first conv block lets us adapt low-level edge and texture filters to lab lighting without disturbing the high-level lane-structure features learned before.

**Comment by Fle-bihh** [comment]
Currently training with following parameters :

Dataset: Team02 Dataset (710 images)

Epochs: 50

Unfreeze encoder_conv1_

Learning rate: original LR / 10

Output_dir : output-20250428-134922

Estimated duration : 50 minutes

EarlyStopping callback monitoring : val_binary_mean_iou

Initial loss: 0.6329 | acc: 0.9696 | IoU: 0.0719

**Comment by Fle-bihh** [comment]
Note: Process to put the model on the car :

1. Use keras_to_onnx.py to convert the newly created keras model to onnx
2. scp the onnx model on the car (ex: scp models/lane_detector_lab_finetuned.onnx $JETSON:/home/hotweels/dev/model_loader/models )
3. use the run_conversion.sh script to convert the onnx to tensorRT (on the car)
4. replace model.engine with the new jetson_model.engine (so that we can use the cross-compiled trt_inference without changing the path its directed to)
5. run trt_inference executable

**Comment by Fle-bihh** [comment]
The dataset did actually contains this scenario, but I trained on it with only the "identifying" layers unfroze, so it did learn to see better the lab lines, but it didn't learn to form lanes on this dataset. So I will do a new fine_tuning with the last model (that recognize the lab lines AND works well on Carla), this time unfreezing only the last decoder block (layers 47–50), which role is to form lanes from previous layers masks.

**Comment by Fle-bihh** [comment]
Currently training with following parameters :

Dataset: Team02 Dataset (710 images)

Epochs: 50

Unfreeze decoder_conv1_1 --> batch_normalization_13 (47 - 50)

Learning rate: original LR / 10

Output_dir : output-20250428-145103

Estimated duration : 50 minutes

EarlyStopping callback monitoring : val_binary_mean_iou

Initial loss: 0.5953 | acc: 0.9700 | IoU: 0.1160

**Comment by Fle-bihh** [comment]
Stopped at epoch 18/50 again.
Final loss: 0.5627 | acc: 0.9722 | IoU: 0.1495

Results are similar on Carla.
Unfortunately, results on the car are still not satisfying. I can see a small improvement in seeing the curves from close distance, but still not enough. At the same time it started seeing lanes that does not exists. Will see now if I need to adjust the dataset, the learning rate or the unfroze layers. But this bad result could mean that the problem is deeper and maybe the best solution would be to train from 0 the same way team02 did as they have very good results on both Carla and lab (Single training on a mix of TUSimple, lab and carla images. The whole with data augmentation)

**Comment by Fle-bihh** [comment]
Will go with Team02's method as time is running out and we have certainty that their method gave good results. Currently downloading the TUSimple dataset and preparing the script to create the mixed dataset and train on it. Augmentation methods were already done from the previous trainings. We will keep our current layers network as I can say it fits this kind of binary segmentation trainings.

**Comment by Fle-bihh** [comment]
Started the new training with the combined TUSimple and team02's dataset. Configuration :
🗒️  Training Summary
  • Dataset        : combined (Team02 + TuSimple)
  • Image size     : (256, 256)
  • Channels       : 1
  • Batch size     : 8
  • Epochs         : 50
  • Learning rate  : 0.001
  • Val split      : 20%
  • Seed           : 31415
  • Output dir     : ./output-20250430-131804

Dataset details :
[TuSimple] Total labeled frames: 3626 (train=2900, val=726)
[TuSimple] Total samples: 3626 (train=2900, val=726)
[Combined] Grand total: 4336 (train=3468, val=868)
📦 Dataset loaded: 4336 samples
   - Train: 3468 samples → 433 steps/epoch
   - Val  : 868 samples → 108 steps

Estimated duration :
6 hours

**Comment by Fle-bihh** [comment]
@t-pereira06 mentioned that team 02's camera doesn't have the same angle as ours (Pointing down while ours points horizontally). When testing this model we should also try playing with the camera angle to see for improvements.

**Comment by Fle-bihh** [comment]
Training was going really well but stopped at epoch 7 because it ran out of data. .repeat was applied on both tuSimple and team02 datasets individually but not on the combined final dataset. I now applied .repeat and .shuffle to it (To make sure it doesnt run out of data and that tusimple and team02 samples's order are mixed ).
Started once again with following config :
  Training Summary
  • Dataset        : combined (Team02 + TuSimple)
  • Image size     : (256, 256)
  • Channels       : 1
  • Batch size     : 8
  • Epochs         : 50
  • Learning rate  : 0.001
  • Val split      : 20%
  • Seed           : 31415
  • Output dir     : ./output-20250430-154802

Estimated duration : 5 hours

**Comment by Fle-bihh** [comment]
Training stopped at epoch 8 due to a stagnation of binary_mean_iou (early callback). Patience was 2 (2 epochs without improving would stop the training). I will restart it with a higher patience (5) And also disabled shuffling and repeating in the individual dataset processing methods since it was already applied on the combined dataset and it was making the training pretty slow.
Starting another training :

🗒️  Training Summary
  • Dataset        : combined (Team02 + TuSimple)
  • Image size     : (256, 256)
  • Channels       : 1
  • Batch size     : 8
  • Epochs         : 50
  • Learning rate  : 0.001
  • Val split      : 20%
  • Seed           : 31415
  • Output dir     : ./output-20250430-165452

Everything seems fine now, no matter if the training early stops, the resulted model should be tested on the car.

**Comment by Fle-bihh** [comment]
May 10 progress

- Ran a training fine_tuning the model obtained with a mix of TUSimple, Lab and Carla images. Using lab curves data. With 10 epochs and data augmentation. Results : No improvements at all. 
- Ran the same training but this time with 50 epoch and an early stopping callback. Results: Early stopped at epoch 26/50, the model forgot everything and couldn't recognized anything anymore on the lab.
- Realized the pre-processing script for the lab curves data (which labels are coordinates of the key points of each lane) was creating wrong masks because using wrong dimensions -> This explained the poor previous results
- Ran a training with the working dataset. Results : Small improvement on curves but start to predict lanes where there is none (intersections of the lab blocks predicted as lane in some scenarios)
- Ran the same training but changed the loss function (from combine_loss to tversky_loss) which fits better when predicting on thin lanes. And unfroze less layers to try and retain the previous results. Results : Very good results on curves ! From distance or close to them, the model can see them clearly almost all the time. Still have this problem where it takes the lab block's intersections for lanes.
- Took a new video with the car to get frames containing these block intersections but no lanes. Associated them with empty label files (because no lanes) and included these new images in the dataset.
- Ran again the fine-tuning with the dataset now containing 222 labeled frames (as before) plus 15 frames with empty mask. Results : The model is the same, seeing very well (but not perfect) the curves, but still interpreting (in certain scenarios) the block intersections as lane.
- Ran again but applied .repeat(3) to negative frames (frames with empty masks) so they don't get "drowned out" (because they represent less then 10% of the dataset). Also augmented the early stopping patience to let the model generalize a bit more (from 2 to 7). Results : Predicting on curves looks even better (but not perfect) and it "imagines" lanes a bit less. 

This is satisfying and seems to be the best we can get with this solution (training with mix of TUSimple / lab / Carla then fine tuning with lab curves). To get the perfect results we want we need to try something else.

**Comment by Fle-bihh** [comment]
Scenario where it sees a lane in the intersection : 
![image](https://github.com/user-attachments/assets/1159fd59-e73d-45fc-a8d8-098047ed5517)

"Satisfying" scenarios : 
![image](https://github.com/user-attachments/assets/ba6ef96a-af4a-4643-aa4f-fcdd80b31fce)
![image](https://github.com/user-attachments/assets/d29592af-b11f-43ed-af5a-18fcb4514666)
![image](https://github.com/user-attachments/assets/82c18f33-150f-43c0-8707-ac981f9521d5)
![image](https://github.com/user-attachments/assets/2fb00663-8854-421f-ac40-65d34fd30599)
![image](https://github.com/user-attachments/assets/e7978d8e-3a4c-4e3c-9ef8-169ac7eb3ed2)

**Comment by Fle-bihh** [comment]
- New solution: Started training from scratch with all the datasets mixed : TUSimple, New version of Team02's dataset (More lab images and more Carla images) and our handmade LabCurves dataset. This time without the first preprocessing layers, which our strong CNNs should already handle. 
- Realized I was training using CPU, so it took me a while but got the training to work with GPU (10 times faster) without errors
- Ran a training without augmentation. Results : Predicts a bit less the straight lines then before, but sees way more the curves (close or far).
- Now running a training with augmentation (rest of the config stays unchanged). Hope for better results

**Comment by Fle-bihh** [comment]
- After applying augmentation, the results on the lab are very good (finally really satisfying) ! I will copy in a safe place this model and the one that works perfectly on Carla. They output the same format so I think it's ok to keep them both and use one for the car and one for simulation. Finally this issue will be closed. I would like to thank my mom and my friends for their support during those months. Also a special shout out to Melo who never stopped believing in me. Siuuuu
